src/vsr/checkpoint_trailer.zig

```diff
@@ -140,269 +143,230 @@ pub fn FreeSetEncodedType(comptime Storage: type) type {
+            assert(trailer.callback == .open);
+            defer assert(trailer.callback == .none);
 
-            assert(set.block_index == 0);
-            assert(!set.grid.?.free_set.opened);
-            assert(set.size_transferred == set.size);
-            assert(set.checksum == vsr.checksum(set.buffer[0..set.size]));
-
-            set.grid.?.free_set.open(.{
-                .encoded = set.buffer[0..set.size],
-                .block_addresses = set.block_addresses[0..set.block_count()],
-            });
+            assert(trailer.block_index == 0);
+            assert(trailer.size_transferred == trailer.size);
+            assert(trailer.checksum == vsr.checksum(trailer.buffer[0..trailer.size]));
 
-            set.grid.?.free_set.opened = true;
-            assert((set.size > 0) == (set.grid.?.free_set.count_acquired() > 0));
-
-            const callback = set.callback.open;
-            set.callback = .none;
-            callback(set);
+            const callback = trailer.callback.open;
+            trailer.callback = .none;
+            callback(trailer);
         }
 
-        /// Checkpoint process is delicate:
-        ///   1. Encode free set.
-        ///   2. Derive the number of blocks required to store the encoding.
-        ///   3. Allocate blocks for the encoding (in the old checkpoint).
-        ///   4. Write the blocks to disk.
-        ///   --------------------------------------------------------------
-        ///   5. Mark currently released blocks as free.
-        ///   6. Release the freshly acquired blocks in the new checkpoint.
-        ///
-        /// This function handles steps 1-4. The caller is responsible for calling
-        /// FreeSet.checkpoint which handles 5 and 6.
-        pub fn checkpoint(set: *Self, callback: *const fn (set: *Self) void) void {
-            assert(set.callback == .none);
-            defer assert(set.callback == .checkpoint);
-
-            {
-                set.grid.?.free_set.include_staging();
-                defer set.grid.?.free_set.exclude_staging();
+        pub fn checkpoint(trailer: *Self, callback: *const fn (trailer: *Self) void) void {
+            assert(trailer.callback == .none);
+            defer assert(trailer.callback == .checkpoint);
 
-                set.size = @as(u32, @intCast(set.grid.?.free_set.encode(set.buffer)));
-                assert(set.size % @sizeOf(FreeSet.Word) == 0);
-                set.size_transferred = 0;
-                set.checksum = vsr.checksum(set.buffer[0..set.size]);
-            }
+            trailer.size_transferred = 0;
+            trailer.checksum = vsr.checksum(trailer.buffer[0..trailer.size]);
 
             {
-                assert(set.grid.?.free_set.count_reservations() == 0);
-                const reservation = set.grid.?.free_set.reserve(set.block_count()).?;
-                defer set.grid.?.free_set.forfeit(reservation);
+                assert(trailer.grid.?.free_set.count_reservations() == 0);
```

Comment:
I wonder whether this is necessary? I think it would be correct to checkpoint a trailer concurrently with something else? Though, it's fine to leave assert as is, just to solidify the current implementation. 

However, let's add the same assert to:

* that place in `grid.zig` where we do "include_staging()" trick. _There_ it is crucialt that there are no outstanding reservations
* inside `include_staging` itself